R/preprocessing.R

```diff
@@ -645,6 +645,138 @@ Load10X_Spatial <- function (
+    meta_data_with_barcodes <- tibble::rownames_to_column(object@meta.data, "barcode")
+    merged_meta_data <- merge(
+      x = meta_data_with_barcodes,
+      y = cell.types,
+      by = "barcode",
+      all.x = TRUE
+    )
+    object@meta.data <- tibble::column_to_rownames(merged_meta_data, "barcode")
+    return(object)
+  } else {
+    return(object)
+  }
+}
+
+#' Load a 10x Genomics Single Cell Experiment into a \code{Seurat} object
+#'
+#' @inheritParams Read10X
+#' @inheritParams SeuratObject::CreateSeuratObject If multiome 10x data the
+#' assay param will not be used. The names of each assay contained in the matrix are used.
+#' @param data.dir Directory containing the H5 file specified by \code{filename}
+#' @param filename Name of H5 file containing the feature barcode matrix
+#' @param to.upper Converts all feature names to upper case. This can provide an
+#' approximate conversion of mouse to human gene names which can be useful in an
+#' explorative analysis. For cross-species comparisons, orthologous genes should
+#' be identified across species and used instead.
+#' @param ... Arguments passed to \code{\link{Read10X_h5}}
+#'
+#' @return A \code{Seurat} object
+#'
+#'
+#' @export
+#' @concept preprocessing
+#'
+#' @examples
+#' \dontrun{
+#' data_dir <- 'path/to/data/directory'
+#' list.files(data_dir) # Should show filtered_feature_bc_matrix.h5
+#' Load10X(data.dir = data_dir)
+#' }
+#'
+Load10X <- function(data.dir, filename = "filtered_feature_bc_matrix.h5",
+                    assay = "RNA", to.upper = FALSE, ...) {
+
+  if (length(data.dir) > 1) {
+    stop("`data.dir` expects a single directory path but received multiple values.")
+  }
+  if (!file.exists(data.dir)) {
+    stop("No such file or directory: '", data.dir, "'")
+  }
+
+
+  filename <- list.files(data.dir, filename, full.names = FALSE, recursive = FALSE)
+  counts.path <- file.path(data.dir, filename)
+  if (!file.exists(counts.path)) {
+    stop("File not found: '", counts.path, "'")
+  }
+
+  counts <- Read10X_h5(counts.path, ...)
+
+  if (to.upper) {
+    counts <- imap(counts, ~{
+      rownames(.x) <- toupper(rownames(.x))
+      .x
+    })
+  }
+
+  if (is.list(counts)) {
+    seurat.list <- lapply(names(counts), function(name) {
+      CreateSeuratObject(
+        counts = counts[[name]],
+        assay = name,
+        project = name
+      )
+    })
+
+    for (i in seq_along(seurat.list)) {
+      if (Assays(seurat.list[[i]]) %in% c("Gene Expression", "RNA")) {
+        seurat.list[[i]] <- Add_10X_CellTypes(data.dir, seurat.list[[i]])
+      }
+    }
+
+    merged.object <- merge(
+      x = seurat.list[[1]],
+      y = seurat.list[-1],
+      add.cell.ids = names(counts),
+      merge.data = FALSE
+    )
+    return(merged.object)
+
+  } else {
+    object <- CreateSeuratObject(counts, assay = assay)
+    if (Assays(object)  %in%  c("Gene Expression", "RNA")) {
+      object <- Add_10X_CellTypes(data.dir, object)
+    }
```

Comment:
I'm not sure that `is.list(counts)` conditional is necessary. Instead, you can just force `counts` to always be a list, and then you can do without the `else` block. 

It looks like there's just one `CellTypes` file per `data.dir`? If that's the case, this could be simplified even more—the call to `Add_10X_CellTypes` could be made on the `merged.object`.